gdnative-derive/src/lib.rs

```diff
@@ -307,6 +299,29 @@ pub fn profiled(meta: TokenStream, input: TokenStream) -> TokenStream {
 ///   }
 ///   ```
 ///
+/// - `async`
+///
+///   Marks the function as async. This is used for functions that aren't `async` themselves, but return `Future`s instead.
+///   This is especially useful for working around Rust's lifetime elision rules, which put the lifetime of `&self` into the
+///   return value for `async fn`s. The `impl Future` syntax instead allows one to explicitly specify a `'static` lifetime,
+///   as required by the async runtime:
+///
+///   ```ignore
+///   // This will NOT compile: Rust assumes that any futures returned by an `async fn` may only live as long as each of its
+///   // arguments, and there is no way to tell it otherwise. As a result, it will emit some cryptic complaints about lifetime.
+///   #[method]
+///   async fn answer(&self) -> i32 {
+///      42
+///   }
```

Comment:
Is this not contradicting the code examples further above, where you have `async fn`s that are apparently? Or is the difference the lack of return value?

Reply:
Technically it isn't, but I see how it can be confusing. As far as the macro is concerned, this declaration *is* valid. The only thing preventing it from compiling is Rust's lifetime elision rules, and those rules can be relaxed down the line such that code like this compile.

I'll try to think of a way to word the examples better.

Reply:
Yep, I think it would help to have a rule for users to know when they can use `async fn` and when not.

Maybe some of these rules could also be detected by the proc-macro itself, to give a better compiler message.

Reply:
There's a problem with macro-based detection though: we still can't emit proper warnings. I don't think we want `async fn (&self)` to be a hard error here, for the reasons above, but any warnings we emit are still through `#[deprecated]`, and as such a user cannot `#[allow]` them precisely without also allowing all the actually deprecated stuff.

Reply:
Re-worded the examples.